lib/postgrex/protocol.ex

```diff
@@ -99,39 +101,76 @@ defmodule Postgrex.Protocol do
       types_lock: nil,
       prepare: prepare,
       messages: [],
-      ssl: ssl?
+      ssl: ssl?,
+      target_server_type: target_server_type,
+      remaining_endpoints: remaining_endpoints_to_attempt
     }
 
-    case connect(host, port, sock_opts ++ @sock_opts, connect_timeout, s) do
-      {:ok, s} -> handshake(s, status)
-      {:error, _} = error -> error
-    end
+    connect_endpoints(host, port, sock_opts ++ @sock_opts, connect_timeout, s, status)
   end
 
-  defp host_and_port(opts) do
+  defp host_and_port_params(opts) do
     port = opts[:port] || 5432
 
     case Keyword.fetch(opts, :socket) do
       {:ok, file} ->
-        {{:local, file}, 0}
+        {{:local, file}, 0, nil}
 
       :error ->
         case Keyword.fetch(opts, :socket_dir) do
           {:ok, dir} ->
-            {{:local, "#{dir}/.s.PGSQL.#{port}"}, 0}
+            {{:local, "#{dir}/.s.PGSQL.#{port}"}, 0, nil}
 
           :error ->
-            case Keyword.fetch(opts, :hostname) do
-              {:ok, hostname} ->
-                {to_charlist(hostname), port}
+            case Keyword.fetch(opts, :endpoints) do
+              {:ok, endpoints} when is_list(endpoints) ->
+                {{host, port}, remaining_endpoints_to_attempt} =
+                  endpoints
+                  |> Enum.map(fn {hostname, port} -> {to_charlist(hostname), port} end)
+                  |> List.pop_at(0)
+
+                {host, port, remaining_endpoints_to_attempt}
+
+              {:ok, _} ->
+                raise ArgumentError, "expected :endpoints to be a list of tuples"
 
               :error ->
-                raise ArgumentError, "expected :hostname, :socket_dir, or :socket to be given"
+                case Keyword.fetch(opts, :hostname) do
+                  {:ok, hostname} ->
+                    {to_charlist(hostname), port, nil}
+
+                  :error ->
+                    raise ArgumentError, "expected :hostname, endpoints, :socket_dir, or :socket to be given"
+                end
             end
         end
     end
   end
 
+  defp connect_endpoints(host, port, sock_opts, timeout, s, %{remaining_endpoints: remaining_endpoints} = status) do
+    case connect_and_proceed(host, port, sock_opts, timeout, s, status) do
+      {:ok, _} = ret ->
+        ret
+
+      {:error, _} when not is_nil(remaining_endpoints) and length(remaining_endpoints) > 0 ->
+        {{host, port}, remaining_endpoints} = List.pop_at(remaining_endpoints, 0)
+        connect_endpoints(host, port, sock_opts, timeout, s, %{status | remaining_endpoints: remaining_endpoints})
```

Comment:
Are you sure you want to update the state? What happens if you are disconnected? You want to go through all endpoints again? To try only what was left?

Reply:
Another option is to not keep this in the state at all... it is used only the first time to find the host and port and then we stick with it until we start a new process.